frontend/src/components/UserUpdateRoles.vue

```diff
@@ -1,210 +1,384 @@
+              >,
+            </span>
+          </span>
         </span>
-      </div>
-    </div>
+      </template>
 
+      <template #item.actions="{ item }">
+        <v-icon small @click="editRoles(item.clientRepresentation)">
+          mdi-pencil
+        </v-icon>
+      </template>
+    </v-data-table>
   </v-card>
 </template>
 
 <script>
 import ClientsRepository from "@/api/ClientsRepository";
 import UsersRepository from "@/api/UsersRepository";
-import UserUpdateSfdsAuth from "@/components/UserUpdateSfdsAuth";
 
 export default {
   name: "UserUpdateRoles",
-  components: { UserUpdateSfdsAuth },
-  props: ['userId'],
+  props: ["userId"],
   data() {
     return {
-      clients: [],
+      dialog: false,
+      headers: [
+        { text: "Application", value: "clientRepresentation.clientId" },
+        { text: "Role", value: "roleArray" },
+        { text: "Last Log In", value: "lastLogin"},
+        { text: "Actions", value: "actions", sortable: false },
+      ],
+      clientWithoutRoles: [],
       selectedClient: null,
       clientRoles: [],
       effectiveClientRoles: [],
-      selectedRoles: []
+      selectedRoles: [],
+      ClientsWithEffectiveRoles: [],
+      rolesLoaded: false,
+      isEdit: false,
     };
   },
   async created() {
-    //TODO error handling
-    await this.getClients();
+   this.loadUserRoles();
   },
   computed: {
     itemsInColumn() {
-      return Math.ceil(this.clientRoles.length / this.numberOfClientRoleColumns);
+      return Math.ceil(
+        this.clientRoles.length / this.numberOfClientRoleColumns
+      );
     },
     numberOfClientRoleColumns() {
-      return (this.clientRoles.length > 10) ? 2 : 1
+      return this.clientRoles.length > 10 ? 2 : 1;
     },
-    hasRoleForManageUserRoles: function() {
+    hasRoleForManageUserRoles: function () {
       const umsClientId = "USER-MANAGEMENT-SERVICE";
       const manageUserRolesName = "manage-user-roles";
 
-      return !!this.$keycloak.tokenParsed.resource_access[umsClientId].roles.includes(manageUserRolesName)
-    }
+      return !!this.$keycloak.tokenParsed.resource_access[
+        umsClientId
+      ].roles.includes(manageUserRolesName);
+    },
   },
   methods: {
-    roleArrayPosition: function(col, item) {
-      return (col - 1) * (this.itemsInColumn) + item - 1;
+    addRoles() {
+      this.selectedClient = null;
+      this.isEdit = false;
+      this.dialog = true;
+    },
+    editRoles: function (client) {
+      this.selectedClient = client;
+      this.isEdit = true;
+      this.dialog = true;
+      this.getUserClientRoles();
     },
-    getClients: function() {
-      return ClientsRepository.get()
-        .then(response => {
-          this.clients = response.data;
+    close() {
+      this.dialog = false;
+    },
+    roleArrayPosition: function (col, item) {
+      return (col - 1) * this.itemsInColumn + item - 1;
+    },
+    addClientRoles: function () {
```

Comment:
My IDE reports `addClientRoles` as an unused property. Many IDE have this sort of scan built-in, but if one is unavailable in may be available via a plug-in or CLI tooling.